dali/kernels/slice/slice_hwc2chw_normalize_gpu.cu

```diff
@@ -0,0 +1,531 @@
+  for (int64_t idx = threadIdx.x + start_x / kStaticChannels, base_x = threadIdx.x;
+       idx < end_x / kStaticChannels; idx += blockDim.x, base_x += blockDim.x) {
+    // TODO(klecki): forceinline device function
+    int64_t out_offset;
+    if constexpr (enable_mirror) {
+      if (sample.flip_x) {
+        int y = idx / sample.W;
+        int x = idx - (int64_t)y * sample.W;
+        int target_x = sample.W - 1 - x;
+        out_offset = (int64_t)y * sample.W + target_x;
+      } else {
+        out_offset = idx;
+      }
+    } else {
+      out_offset = idx;
+    }
+
+    #pragma unroll kStaticChannels
+    for (int c = 0; c < kStaticChannels; c++) {
+      // the kStaticChannels == input_C
+      float fpin = prologue_tile[base_x * sample.input_C + c];
+      float fpout = fmaf(fpin, norm_mul[c], norm_add[c]);
+      sample.out[c * sample.H * sample.W + out_offset] = ConvertSat<Out>(fpout);
+    }
+
+    if constexpr (enable_pad) {
+      for (int c = kStaticChannels; c < sample.C; c++) {
+        sample.out[c * sample.H * sample.W + out_offset] = fill_values[c];
+      }
+    }
+  }
+}
+
+/**
+ * @brief Slice Hwc2Chw Normalize Mirror-x Pad-channel kernel
+ * This kernel supports cropping in x-coordinate.
+ * It extends the input loop, by utilizing the (unaligned prologue, aligned main loop, epilogue)
+ * pattern in a row-by-row loop.
+ * Indexing is based on the output coordinates, specifically we read rows for coordinate X
+ * between 0 and output H. (The samples are shifted so they always start from 0 in X).
+ */
+template <typename Out, typename In, bool enable_mirror, bool enable_pad, int kBlockSize,
+          int kStaticChannels>
+__global__ void SliceHwc2ChwNormalize(const Hwc2ChwSampleDesc<Out, In> *samples,
+                                      uint32_t num_samples) {
+  // TODO(klecki): generalize for wider input types
+  static_assert(std::is_same<In, uint8_t>::value, "Only uint8_t supported as input");
+
+  int sample_idx = FindSampleIdx(samples, num_samples);
+
+  const auto sample = samples[sample_idx];
+  int64_t start_x = (blockIdx.x - sample.first_block) * kBlockSize;
+  int64_t end_x = ::min(start_x + kBlockSize, sample.sample_size);
+
+  __shared__ float tile[kBlockSize + 32 * 4];
+
+  // Preload the norm values so they are accessed via registers and not from gmem via pointer.
+  float norm_mul[kStaticChannels], norm_add[kStaticChannels];
+
+  #pragma unroll kStaticChannels
+  for (int c = 0; c < kStaticChannels; c++) {
+    norm_mul[c] = sample.norm_mul[c];
+    norm_add[c] = sample.norm_add[c];
+  }
+
+  // Strides use the input number of channels without the padding
+  int in_stride = sample.input_W * sample.input_C;
+  int out_stride = sample.W * sample.input_C;
+
+  // The rows we start and end with, we are indexed by output coordinates
+  int y_start = start_x / out_stride;
+  int y_end = end_x / out_stride + 1;
+
+  float *tile_row = tile;
+
+  for (int y = y_start; y < y_end; y++) {
```

Comment:
This looks suspicious with the inner block-strided loop. If the slice is narrow (< 4*blockDim.x), then many threads will do nothing.

Reply:
We include the channel dimension here, so the narrowest slice that utilizes the whole block is about 44 pixels. It's not that narrow.